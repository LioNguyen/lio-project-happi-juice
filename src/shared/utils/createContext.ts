import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
