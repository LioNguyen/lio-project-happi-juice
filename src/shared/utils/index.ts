export * from './lazyLoad'
export * from './tailwind'
