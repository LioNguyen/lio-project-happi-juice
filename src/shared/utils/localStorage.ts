/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
