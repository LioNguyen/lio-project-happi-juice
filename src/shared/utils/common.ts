import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
