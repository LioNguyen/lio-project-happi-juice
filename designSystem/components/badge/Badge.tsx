import * as React from 'react'
import { cva, type VariantProps } from 'class-variance-authority'

import { cn } from '@/shared/utils'

const badgeVariants = cva(
  'inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2',
  {
    variants: {
      variant: {
        default:
          'border-transparent bg-primary text-text-secondary shadow hover:bg-primary/80',
        secondary:
          'border-transparent bg-background-secondary text-text-primary hover:bg-secondary/80',
        outline: 'text-text-primary',
        active:
          'border-badge-active-border bg-badge-active-background text-badge-active-foreground hover:bg-badge-active-background/50',
        inactive:
          'border-badge-inactive-border bg-badge-inactive-background text-badge-inactive-foreground hover:bg-badge-inactive-background/50',
        expired:
          'border-badge-expired-border bg-badge-expired-background text-badge-expired-foreground hover:bg-badge-expired-background/50',
      },
    },
    defaultVariants: {
      variant: 'default',
    },
  },
)

interface BadgeProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof badgeVariants> {}

function Badge({ className, variant, ...props }: BadgeProps) {
  return (
    <div className={cn(badgeVariants({ variant }), className)} {...props} />
  )
}

export { Badge, badgeVariants }
export type { BadgeProps }
