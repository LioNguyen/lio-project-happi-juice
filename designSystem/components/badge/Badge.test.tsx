import { screen } from "@testing-library/react";
import { VariantProps } from "class-variance-authority";
import { ComponentPropsWithRef } from "react";
import { describe, expect, it } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import { Badge, badgeVariants } from "./Badge";

type BadgePropsWithRef = ComponentPropsWithRef<typeof Badge>;
type BadgeVariantProps = VariantProps<typeof badgeVariants>;
type BadgeVariant = NonNullable<BadgeVariantProps["variant"]>;

const defaultProps: Partial<BadgePropsWithRef> = {
  variant: "default",
};

const renderComponent = (props: Partial<BadgePropsWithRef> = {}) => {
  return renderWithTheme(<Badge {...defaultProps} {...props} />);
};

describe("Badge Component", () => {
  const defaultClass =
    "inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2";

  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  // Test render
  it("should render badge with default props", () => {
    const { container } = renderComponent({ children: "Status" });
    const badge = screen.getByText("Status");

    expect(badge).toBeInTheDocument();
    expect(container.firstChild).toHaveClass(defaultClass);
  });

  // Test className custom
  it("should merge custom className with default classes", () => {
    const { container } = renderComponent({
      className: "custom-class",
      children: "Status",
    });

    const badge = container.firstChild;
    expect(badge).toHaveClass("custom-class");
    expect(badge).toHaveClass(defaultClass);
  });

  // Test HTML attributes
  it("should forward HTML attributes", () => {
    const { container } = renderComponent({
      children: "Status",
      "aria-label": "Status Badge",
    });

    expect(container.firstChild).toHaveAttribute("aria-label", "Status Badge");
  });

  // Test variants
  describe("Variants", () => {
    const variants: BadgeVariant[] = ["default", "secondary", "outline", "active", "inactive", "expired"] as const;

    variants.forEach(variant => {
      it(`should render ${variant} variant correctly`, () => {
        const { container } = renderComponent({
          variant,
          children: "Status",
        });

        const badge = container.firstChild;

        switch (variant) {
          case "default":
            expect(badge).toHaveClass("border-transparent bg-primary text-text-secondary shadow hover:bg-primary/80");
            break;
          case "secondary":
            expect(badge).toHaveClass("border-transparent bg-secondary text-text-primary hover:bg-secondary/80");
            break;
          case "outline":
            expect(badge).toHaveClass("text-text-primary");
            break;
          case "active":
            expect(badge).toHaveClass(
              "border-badge-active-border bg-badge-active-background text-badge-active-foreground hover:bg-badge-active-background/50",
            );
            break;
          case "inactive":
            expect(badge).toHaveClass(
              "border-badge-inactive-border bg-badge-inactive-background text-badge-inactive-foreground hover:bg-badge-inactive-background/50",
            );
            break;
          case "expired":
            expect(badge).toHaveClass(
              "border-badge-expired-border bg-badge-expired-background text-badge-expired-foreground hover:bg-badge-expired-background/50",
            );
            break;
        }
      });
    });
  });

  // Test accessibility
  it("should have appropriate ARIA attributes", () => {
    const { container } = renderComponent({
      children: "Status",
      "aria-label": "Status Badge",
      role: "status",
    });

    const badge = container.firstChild;
    expect(badge).toHaveAttribute("role", "status");
    expect(badge).toHaveAttribute("aria-label", "Status Badge");
  });
});
