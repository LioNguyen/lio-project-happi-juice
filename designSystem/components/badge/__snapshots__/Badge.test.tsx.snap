// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge Component > should match snapshot with default props 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-text-secondary shadow hover:bg-primary/80"
/>
`;
